File: Java/169.majority-element.java
public class Solution {
	public int majorityElement(int[] nums) {
    int[] bit = new int[32];
    for (int num: nums)
        for (int i=0; i<32; i++)
            if ((num>>(31-i) & 1) == 1)
                bit[i]++;
    int ret=0;
    for (int i=0; i<32; i++) {
        bit[i]=bit[i]>nums.length/2?1:0;
        ret += bit[i]*(1<<(31-i));
    }
    return ret;
}

}

File: 2026/169.majority-element.js
/*
 * @lc app=leetcode id=169 lang=javascript
 *
 * [169] Majority Element
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function (nums) {
  let hash = {}, max = 0, val = nums[0];

  for (let [num, i] of nums) {
    hash[num] ? hash[num] += 1 : hash[num] = 1;

    if (hash[num] > max) {
      max = hash[num];
      val = num;
    }
  }

  return val;
};

// O(1) space solution: 摩尔投票算法：核心理念为 票数正负抵消
var majorityElement = function (nums) {
  let count = 0, candidate = null;

  for (let num of nums) {
    if (count === 0) {
      candidate = num;
    }
    count += (num === candidate) ? 1 : -1;
  }

  return candidate;
}


// @lc code=end
